File: AndroidArchitecture/mvvm-databinding/src/main/java/com/codearms/maoqiqi/app/data/source/room/TasksRoomDataSource.java
package com.codearms.maoqiqi.app.data.source.room;

import android.content.Context;
import android.support.annotation.VisibleForTesting;

import com.codearms.maoqiqi.app.data.TaskBean;
import com.codearms.maoqiqi.app.data.source.TasksDataSource;
import com.codearms.maoqiqi.app.utils.AppExecutors;

import java.util.List;

/**
 * Use room to create the sqlite database as the data source.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 15:18
 */
public class TasksRoomDataSource implements TasksDataSource {

    private static volatile TasksRoomDataSource INSTANCE;

    private final TasksDAO dao;
    private final AppExecutors appExecutors;

    private TasksRoomDataSource(Context context, AppExecutors appExecutors) {
        this.dao = TaskDataBase.getInstance(context).tasksDAO();
        this.appExecutors = appExecutors;
    }

    public static TasksRoomDataSource getInstance(Context context, AppExecutors appExecutors) {
        if (INSTANCE == null) {
            synchronized (TasksRoomDataSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new TasksRoomDataSource(context, appExecutors);
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    static void clearInstance() {
        INSTANCE = null;
    }

    @Override
    public void loadTasks(final LoadTasksCallBack callBack) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                final List<TaskBean> list = dao.loadTasks();
                appExecutors.getMainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        if (list.isEmpty()) {
                            callBack.onDataNotAvailable();
                        } else {
                            callBack.onTasksLoaded(list);
                        }
                    }
                });
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void getTask(final String taskId, final GetTaskCallBack callBack) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                final TaskBean taskBean = dao.getTaskById(taskId);
                appExecutors.getMainThread().execute(new Runnable() {
                    @Override
                    public void run() {
                        if (taskBean == null) {
                            callBack.onDataNotAvailable();
                        } else {
                            callBack.onTaskLoaded(taskBean);
                        }
                    }
                });
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void clearCompletedTasks() {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.deleteCompletedTasks();
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void refreshTasks() {

    }

    @Override
    public void addTask(final TaskBean taskBean) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.addTask(taskBean);
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void updateTask(final TaskBean taskBean) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.updateTask(taskBean);
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void completeTask(TaskBean completedTaskBean) {
        completeTask(completedTaskBean.getId());
    }

    @Override
    public void completeTask(final String taskId) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.updateCompleted(taskId, true);
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void activateTask(TaskBean activeTaskBean) {
        activateTask(activeTaskBean.getId());
    }

    @Override
    public void activateTask(final String taskId) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.updateCompleted(taskId, false);
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void deleteTask(final String taskId) {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.deleteTaskById(taskId);
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }

    @Override
    public void deleteAllTasks() {
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                dao.deleteTasks();
            }
        };
        appExecutors.getDiskIO().execute(runnable);
    }
}

File: AndroidArchitecture/mvp-kotlin/src/main/java/com/codearms/maoqiqi/app/data/source/room/TasksRoomDataSource.kt
package com.codearms.maoqiqi.app.data.source.room

import android.content.Context
import androidx.annotation.VisibleForTesting

import com.codearms.maoqiqi.app.data.TaskBean
import com.codearms.maoqiqi.app.data.source.TasksDataSource
import com.codearms.maoqiqi.app.utils.AppExecutors

/**
 * Use room to create the sqlite database as the data source.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 15:18
 */
class TasksRoomDataSource private constructor(
        context: Context,
        private val appExecutors: AppExecutors
) : TasksDataSource {

    private val dao: TasksDAO = TaskDataBase.getInstance(context).tasksDAO()

    override fun loadTasks(callBack: TasksDataSource.LoadTasksCallBack) {
        appExecutors.diskIO.execute {
            val list = dao.loadTasks()
            appExecutors.mainThread.execute {
                if (list.isEmpty()) {
                    callBack.onDataNotAvailable()
                } else {
                    callBack.onTasksLoaded(list)
                }
            }
        }
    }

    override fun getTask(taskId: String, callBack: TasksDataSource.GetTaskCallBack) {
        appExecutors.diskIO.execute {
            val taskBean = dao.getTaskById(taskId)
            appExecutors.mainThread.execute {
                if (taskBean == null) {
                    callBack.onDataNotAvailable()
                } else {
                    callBack.onTaskLoaded(taskBean)
                }
            }
        }
    }

    override fun clearCompletedTasks() {
        appExecutors.diskIO.execute { dao.deleteCompletedTasks() }
    }

    override fun refreshTasks() {

    }

    override fun addTask(taskBean: TaskBean) {
        appExecutors.diskIO.execute { dao.addTask(taskBean) }
    }

    override fun updateTask(taskBean: TaskBean) {
        appExecutors.diskIO.execute { dao.updateTask(taskBean) }
    }

    override fun completeTask(completedTaskBean: TaskBean) {
        completeTask(completedTaskBean.id)
    }

    override fun completeTask(taskId: String) {
        appExecutors.diskIO.execute { dao.updateCompleted(taskId, true) }
    }

    override fun activateTask(activeTaskBean: TaskBean) {
        activateTask(activeTaskBean.id)
    }

    override fun activateTask(taskId: String) {
        appExecutors.diskIO.execute { dao.updateCompleted(taskId, false) }
    }

    override fun deleteTask(taskId: String) {
        appExecutors.diskIO.execute { dao.deleteTaskById(taskId) }
    }

    override fun deleteAllTasks() {
        appExecutors.diskIO.execute { dao.deleteTasks() }
    }

    companion object {
        @Volatile
        private var INSTANCE: TasksRoomDataSource? = null

        @JvmStatic
        fun getInstance(context: Context, appExecutors: AppExecutors): TasksRoomDataSource = INSTANCE
                ?: synchronized(TasksRoomDataSource::class.java) {
                    INSTANCE ?: TasksRoomDataSource(context, appExecutors).also { INSTANCE = it }
                }

        @VisibleForTesting
        @JvmStatic
        internal fun clearInstance() {
            INSTANCE = null
        }
    }
}